rootfs/etc/nginx/lua/balancer/chashsubset.lua

```diff
@@ -0,0 +1,92 @@
+-- Consistent hashing to a subset of nodes. Instead of returning the same node
+-- always, we return the same subset always.
+
+local resty_chash = require("resty.chash")
+local util = require("util")
+
+local _M = { name = "chashsubset" }
+
+local function build_nodeset_map(backend)
+  local nodes = {}
+  local nodeset_map = {}
+  local node_sets = {}
+  local subset_size = 3
+
+  if backend["upstream-hash-subset-size"] == 0 then
+    subset_size = tonumber(backend["upstream-hash-subset-size"])
+  end
+
+  for _, endpoint in pairs(backend.endpoints) do
+    local endpoint_string = endpoint.address .. ":" .. endpoint.port
+    table.insert(nodes, endpoint_string)
+  end
+
+  local set_count = math.ceil(#nodes/subset_size)
+  local node_count = set_count * subset_size
+
+  -- if we don't have enough nodes, we use reuse nodes in the last set to
+  -- keep the same same on all of them.
+  if #nodes < node_count then
+    local j = 1
+    for _ = #nodes+1, node_count do
+      table.insert(nodes, nodes[j])
+      j = j+1
+    end
+  end
+
+  local k = 1
+  for i = 1, set_count do
+    local nodeset = {}
+    local nodeset_id = "set" .. tostring(i)
+    for _ = 1, subset_size do
+      table.insert(nodeset, nodes[k])
+      k = k+1
+    end
+    node_sets[nodeset_id] = nodeset
+    nodeset_map[nodeset_id] = 1
+  end
+
+  return nodeset_map, node_sets
+end
+
+function _M.new(self, backend)
+  local nodeset_map, node_sets = build_nodeset_map(backend)
+
+  local o = {
+    instance = resty_chash:new(nodeset_map),
+    hash_by = backend["upstream-hash-by"],
+    node_sets = node_sets,
+    current_endpoints = backend.endpoints
+  }
+  setmetatable(o, self)
+  self.__index = self
+  return o
+end
+
+math.randomseed(os.time())
```

Comment:
We should not do seeding in individual modules, it should be done once only when Nginx worker process starts. Otherwise it can result in weakened randomness.